### PR TITLE
fix: import map with trailing slash

### LINKED ISSUE #141: Better import map detection
**Describe the bug**

Scenario 1:

Following example from https://github.com/WICG/import-maps#packages-via-trailing-slashes with the trailing slash entry at the bottom.

import_map.json
```json
{
  "imports": {
    "deno_std:log": "https://deno.land/std@v0.40.0/log/mod.ts",
    "deno_std:log/": "https://deno.land/std@v0.40.0/log/",
  }
}
```

The extension is ignoring `"deno_std:log/"`.

Screen Shot 2020-04-14 at 3 32 10 PM

Scenario 2:

Simply flipping the order.

import_map.json
```json
{
  "imports": {
    "deno_std:log/": "https://deno.land/std@v0.40.0/log/",
    "deno_std:log": "https://deno.land/std@v0.40.0/log/mod.ts",
  }
}
```

The extension has no problem detecting both.

Screen Shot 2020-04-14 at 3 41 08 PM

Both of these scenarios have no issue with `deno --importmap`.

**Expected behavior**

Spec wise, the extension should detect the trailing slash in import map regardless of order of entry.

**Versions**

vscode: 1.44.0
deno: 0.40.0
extension: 3.5.1

### PR DESCRIPTION
**Pull request by shian15810**
Fixes https://github.com/axetroy/vscode-deno/issues/141.

### CODE REVIEW

**Comment by shian15810** [comment]
@axetroy I have no authority over this coveralls failing the test. Please help, thanks.

**Comment by shian15810** [comment]
@axetroy Thanks! Test is ok now.

File: core/import_map.ts [hunk 1]
```diff
@@ -54,6 +54,17 @@ export class ImportMap implements ImportMapInterface {
       }
     }
 
+    const imports = Object.entries(importMap.imports)
+      .filter(
+        ([key, val]) =>
+          (key.endsWith("/") && val.endsWith("/")) ||
+          (!key.endsWith("/") && !val.endsWith("/"))
```
Comment (axetroy) [inline core/import_map.ts:61]: nice work!

But this is just filtering out invalid rules.

Could you add the test to make sure the rules like this work fine?

```json
{
  "imports": {
    "moment": "/node_modules/moment/src/moment.js",
    "moment/": "/node_modules/moment/src/",
    "lodash": "/node_modules/lodash-es/lodash.js",
    "lodash/": "/node_modules/lodash-es/"
  }
}
```

This requires adding test cases to `module_resolver.test.ts`

edit: 

ignore the last message. just need more test cases for `import_map.test.ts`

```ts
  expect(
    ImportMap.create(validImportMapFilepath).resolveModule("demo/mod.ts")
  ).toEqual("https://example.com/demo/mod.ts");
```

File: core/import_map.ts [hunk 2]
```diff
@@ -54,6 +54,17 @@ export class ImportMap implements ImportMapInterface {
...
+      )
+      .sort(([key1], [key2]) => key2.indexOf("/") - key1.indexOf("/"))
```
Comment (shian15810) [inline core/import_map.ts]: I will change this to `lastIndexOf`. `indexOf` might create some issues if the key has forward `/` or in the middle.

**Comment by shian15810** [comment]
Just FYI, I'm experiencing this issue a lot: https://github.com/kulshekhar/ts-jest/issues/1506

Should be fixed in next release of jest, but not yet as for this moment.

ATM, I'm using `jest --no-cache` to workaround this issue.

File: core/import_map.ts (hunk 2 above)
Comment (shian15810) [reply core/import_map.ts]: Done!

File: core/import_map.ts (hunk 1 above)
Comment (shian15810) [reply core/import_map.ts:61]: Done!

**Comment by shian15810** [comment]
@axetroy Should be ok.

**Comment by shian15810** [comment]
@axetroy Should be ok. This should be final.

**Review by axetroy — APPROVED** [review]
great work & Thx.

**Comment by axetroy** [comment]
@shian15810 This has land in v3.6.0

**Comment by shian15810** [comment]
@axetroy Just tested it, no issue! Thank you.

**Comment by btrauma8** [comment]
i have 3.6.0, but this isn't working for me when trying to setup import maps for local directories.

am i doing something wrong?
image below should show what's going on...

sshot

**Comment by shian15810** [comment]
@btrauma8

I have no problem using absolute and relative path in `import_map.json`.
However, if I specify the `file:` protocol, I get the same error like you.

You can try relative path like `./my-lib` or absolute path like `c:/junk/deno/my-lib`.

I think you should raise another issue regarding the extension having problem with `file:` protocol in `import_map.json`.

**Comment by btrauma8** [comment]
ahhhh....
imports: { "my-lib/": "./my-lib/" } worked for both deno and vscode plugin
c:/junk... only worked for vscode plugin, not deno.

for some reason i thought the importmap had to use file: protocol if pointing to local files,
but apparently not. relative offset is nicer anyway, so, thanks!

now i can get back the tsconfig.json baseUrl functionality that i miss

**Comment by shian15810** [comment]
@btrauma8 I think you definitely should raise an issue about this problem.

In any way, this extension should comply to specs and deno standards.

**Comment by btrauma8** [comment]
done. https://github.com/axetroy/vscode-deno/issues/146

**Pull request merged**